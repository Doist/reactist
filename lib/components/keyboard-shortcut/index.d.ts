import { KeyboardShortcut } from './keyboard-shortcut';
export default KeyboardShortcut;
