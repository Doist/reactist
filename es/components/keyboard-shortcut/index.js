import { KeyboardShortcut } from './keyboard-shortcut.js'
export { KeyboardShortcut as default } from './keyboard-shortcut.js'
//# sourceMappingURL=index.js.map
