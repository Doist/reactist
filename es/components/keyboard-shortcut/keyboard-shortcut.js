import {
    objectWithoutProperties as _objectWithoutProperties,
    objectSpread2 as _objectSpread2,
} from '../../_virtual/_rollupPluginBabelHelpers.js'
import * as React from 'react'
import classNames from 'classnames'

const _excluded = ['children', 'className', 'translateKey', 'isMac']
// Support for setting up how to translate modifiers globally.
//

let globalTranslateKey = (key) => key

KeyboardShortcut.setTranslateKey = (tr) => {
    globalTranslateKey = tr
}

function translateKeyMac(key) {
    switch (key.toLowerCase()) {
        case 'cmd':
        case 'mod':
            return '⌘'

        case 'control':
        case 'ctrl':
            return '⌃'

        case 'alt':
            return '⌥'

        case 'shift':
            return '⇧'

        case 'space':
            return '␣'

        default:
            return key
    }
} //
// Some helpers
//

function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()
}

function hasModifiers(str) {
    return /\b(mod|cmd|ctrl|control|alt|shift)\b/i.test(str)
}

function isSpecialKey(str) {
    return /^(mod|cmd|ctrl|control|alt|shift|space|super)$/i.test(str)
}

function parseKeys(shortcut, isMac, translateKey) {
    const t = isMac ? translateKeyMac : translateKey

    const _hasModifiers = hasModifiers(shortcut)

    function mapIndividualKey(str) {
        if (isSpecialKey(str)) {
            return capitalize(t(str))
        }

        if (_hasModifiers && str.length === 1) {
            return str.toUpperCase()
        }

        return str
    }

    if (!isMac) {
        shortcut = shortcut.replace(/\b(mod|cmd)\b/i, 'ctrl')
    }

    return shortcut.split(/\s*\+\s*/).map(mapIndividualKey)
}

function KeyboardShortcut(_ref) {
    var _navigator$platform$t, _navigator$platform

    let {
            children,
            className,
            translateKey = globalTranslateKey,
            isMac = (_navigator$platform$t =
                (_navigator$platform = navigator.platform) == null
                    ? void 0
                    : _navigator$platform.toUpperCase().includes('MAC')) != null
                ? _navigator$platform$t
                : false,
        } = _ref,
        props = _objectWithoutProperties(_ref, _excluded)

    const shortcuts = typeof children === 'string' ? [children] : children
    return /*#__PURE__*/ React.createElement(
        'span',
        _objectSpread2(
            {
                className: classNames('reactist_keyboard_shortcut', className, {
                    'reactist_keyboard_shortcut--macos': isMac,
                }),
            },
            props,
        ),
        shortcuts.map((shortcut, i) =>
            /*#__PURE__*/ React.createElement(
                React.Fragment,
                {
                    key: i,
                },
                i === 0 ? null : ', ',
                /*#__PURE__*/ React.createElement(
                    'kbd',
                    null,
                    parseKeys(shortcut, isMac, translateKey).map((key, j) =>
                        /*#__PURE__*/ React.createElement(
                            'kbd',
                            {
                                key: j,
                            },
                            key,
                        ),
                    ),
                ),
            ),
        ),
    )
}

export { KeyboardShortcut }
//# sourceMappingURL=keyboard-shortcut.js.map
