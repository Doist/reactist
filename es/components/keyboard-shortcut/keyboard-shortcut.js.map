{"version":3,"file":"keyboard-shortcut.js","sources":["../../../src/components/keyboard-shortcut/keyboard-shortcut.tsx"],"sourcesContent":["import * as React from 'react'\nimport classNames from 'classnames'\n\n//\n// Support for setting up how to translate modifiers globally.\n//\n\nlet globalTranslateKey = (key: string) => key\n\ntype TranslateKey = typeof globalTranslateKey\n\nKeyboardShortcut.setTranslateKey = (tr: TranslateKey) => {\n    globalTranslateKey = tr\n}\n\nfunction translateKeyMac(key: string) {\n    switch (key.toLowerCase()) {\n        case 'cmd':\n        case 'mod':\n            return '⌘'\n        case 'control':\n        case 'ctrl':\n            return '⌃'\n        case 'alt':\n            return '⌥'\n        case 'shift':\n            return '⇧'\n        case 'space':\n            return '␣'\n        default:\n            return key\n    }\n}\n\n//\n// Some helpers\n//\n\nfunction capitalize(str: string) {\n    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()\n}\n\nfunction hasModifiers(str: string) {\n    return /\\b(mod|cmd|ctrl|control|alt|shift)\\b/i.test(str)\n}\n\nfunction isSpecialKey(str: string) {\n    return /^(mod|cmd|ctrl|control|alt|shift|space|super)$/i.test(str)\n}\n\nfunction parseKeys(shortcut: string, isMac: boolean, translateKey: TranslateKey) {\n    const t = isMac ? translateKeyMac : translateKey\n    const _hasModifiers = hasModifiers(shortcut)\n\n    function mapIndividualKey(str: string) {\n        if (isSpecialKey(str)) {\n            return capitalize(t(str))\n        }\n        if (_hasModifiers && str.length === 1) {\n            return str.toUpperCase()\n        }\n        return str\n    }\n\n    if (!isMac) {\n        shortcut = shortcut.replace(/\\b(mod|cmd)\\b/i, 'ctrl')\n    }\n\n    return shortcut.split(/\\s*\\+\\s*/).map(mapIndividualKey)\n}\n\n//\n// The KeyboardShortcut component\n//\n\ntype NativeSpanProps = React.DetailedHTMLProps<\n    React.HTMLAttributes<HTMLSpanElement>,\n    HTMLSpanElement\n>\n\ntype Props = Omit<NativeSpanProps, 'children'> & {\n    /**\n     * The shortcut to be represented as markup. It supports an intuitive syntax where you can\n     * combine modifiers (cmd, ctrl, shift, alt) with single keys all concatenated with plus signs.\n     * You can also pass various shortcuts as an array, which will be depicted separated by commas.\n     */\n    children: string | string[]\n    /**\n     * A function that allows you to change how some key names are represented. This may be useful,\n     * for instance, to translate modifier names that are expressed differently in other languages\n     * (e.g. `Ctrl` is named `Strg` in German).\n     *\n     * It defaults to a global version that leaves the key as is. You can pass your version as a\n     * prop, or you can also set your own version of this global default one, so you don't need to\n     * pass your own on each invocation of this component.\n     *\n     * ```js\n     * import { KeyboardShortcut } from '@doist/reactist'\n     * KeyboardShortcut.setTranslateKey = key => { ... }\n     * ```\n     *\n     * Note: When the component detects the macOS operating system it bypasses key translation for\n     * most modifiers and uses macOS-specific symbols. See the `isMac` prop for details.\n     */\n    translateKey?: TranslateKey\n    /**\n     * This prop is not meant to be passed. The component will automatically initialize it to `true`\n     * if it detects that the current browser / operating system is on macOS, in which case modifier\n     * keys are represented using macOS' notation (e.g. ⌘ ⌃ ⌥ ⇧).\n     *\n     * Though it is discouraged, if you don't want this special treatment in macOS, you can pass\n     * `isMac={false}` in all invocations of this component.\n     */\n    isMac?: boolean\n}\n\nfunction KeyboardShortcut({\n    children,\n    className,\n    translateKey = globalTranslateKey,\n    isMac = navigator.platform?.toUpperCase().includes('MAC') ?? false,\n    ...props\n}: Props) {\n    const shortcuts = typeof children === 'string' ? [children] : children\n    return (\n        <span\n            className={classNames('reactist_keyboard_shortcut', className, {\n                'reactist_keyboard_shortcut--macos': isMac,\n            })}\n            {...props}\n        >\n            {shortcuts.map((shortcut, i) => (\n                <React.Fragment key={i}>\n                    {i === 0 ? null : ', '}\n                    <kbd>\n                        {parseKeys(shortcut, isMac, translateKey).map((key, j) => (\n                            <kbd key={j}>{key}</kbd>\n                        ))}\n                    </kbd>\n                </React.Fragment>\n            ))}\n        </span>\n    )\n}\n\nexport { KeyboardShortcut }\n"],"names":["globalTranslateKey","key","KeyboardShortcut","setTranslateKey","tr","translateKeyMac","toLowerCase","capitalize","str","charAt","toUpperCase","slice","hasModifiers","test","isSpecialKey","parseKeys","shortcut","isMac","translateKey","t","_hasModifiers","mapIndividualKey","length","replace","split","map","children","className","navigator","platform","includes","props","shortcuts","React","_objectSpread","classNames","i","createElement","Fragment","j"],"mappings":";;;;;AAIA;AACA;;AAEA,IAAIA,kBAAkB,GAAIC,GAAD,IAAiBA,GAA1C,CAAA;;AAIAC,gBAAgB,CAACC,eAAjB,GAAoCC,EAAD,IAAqB;AACpDJ,EAAAA,kBAAkB,GAAGI,EAArB,CAAA;AACH,CAFD,CAAA;;AAIA,SAASC,eAAT,CAAyBJ,GAAzB,EAAoC;EAChC,QAAQA,GAAG,CAACK,WAAJ,EAAR;AACI,IAAA,KAAK,KAAL,CAAA;AACA,IAAA,KAAK,KAAL;AACI,MAAA,OAAO,GAAP,CAAA;;AACJ,IAAA,KAAK,SAAL,CAAA;AACA,IAAA,KAAK,MAAL;AACI,MAAA,OAAO,GAAP,CAAA;;AACJ,IAAA,KAAK,KAAL;AACI,MAAA,OAAO,GAAP,CAAA;;AACJ,IAAA,KAAK,OAAL;AACI,MAAA,OAAO,GAAP,CAAA;;AACJ,IAAA,KAAK,OAAL;AACI,MAAA,OAAO,GAAP,CAAA;;AACJ,IAAA;AACI,MAAA,OAAOL,GAAP,CAAA;AAdR,GAAA;AAgBH;AAGD;AACA;;;AAEA,SAASM,UAAT,CAAoBC,GAApB,EAA+B;AAC3B,EAAA,OAAOA,GAAG,CAACC,MAAJ,CAAW,CAAX,CAAcC,CAAAA,WAAd,EAA8BF,GAAAA,GAAG,CAACG,KAAJ,CAAU,CAAV,CAAA,CAAaL,WAAb,EAArC,CAAA;AACH,CAAA;;AAED,SAASM,YAAT,CAAsBJ,GAAtB,EAAiC;AAC7B,EAAA,OAAO,uCAAwCK,CAAAA,IAAxC,CAA6CL,GAA7C,CAAP,CAAA;AACH,CAAA;;AAED,SAASM,YAAT,CAAsBN,GAAtB,EAAiC;AAC7B,EAAA,OAAO,iDAAkDK,CAAAA,IAAlD,CAAuDL,GAAvD,CAAP,CAAA;AACH,CAAA;;AAED,SAASO,SAAT,CAAmBC,QAAnB,EAAqCC,KAArC,EAAqDC,YAArD,EAA+E;AAC3E,EAAA,MAAMC,CAAC,GAAGF,KAAK,GAAGZ,eAAH,GAAqBa,YAApC,CAAA;;AACA,EAAA,MAAME,aAAa,GAAGR,YAAY,CAACI,QAAD,CAAlC,CAAA;;EAEA,SAASK,gBAAT,CAA0Bb,GAA1B,EAAqC;AACjC,IAAA,IAAIM,YAAY,CAACN,GAAD,CAAhB,EAAuB;AACnB,MAAA,OAAOD,UAAU,CAACY,CAAC,CAACX,GAAD,CAAF,CAAjB,CAAA;AACH,KAAA;;AACD,IAAA,IAAIY,aAAa,IAAIZ,GAAG,CAACc,MAAJ,KAAe,CAApC,EAAuC;MACnC,OAAOd,GAAG,CAACE,WAAJ,EAAP,CAAA;AACH,KAAA;;AACD,IAAA,OAAOF,GAAP,CAAA;AACH,GAAA;;EAED,IAAI,CAACS,KAAL,EAAY;IACRD,QAAQ,GAAGA,QAAQ,CAACO,OAAT,CAAiB,gBAAjB,EAAmC,MAAnC,CAAX,CAAA;AACH,GAAA;;EAED,OAAOP,QAAQ,CAACQ,KAAT,CAAe,UAAf,CAA2BC,CAAAA,GAA3B,CAA+BJ,gBAA/B,CAAP,CAAA;AACH,CAAA;;AA+CD,SAASnB,gBAAT,CAMQ,IAAA,EAAA;AAAA,EAAA,IAAA,qBAAA,EAAA,mBAAA,CAAA;;EAAA,IANkB;IACtBwB,QADsB;IAEtBC,SAFsB;AAGtBT,IAAAA,YAAY,GAAGlB,kBAHO;AAItBiB,IAAAA,KAAK,GAAGW,CAAAA,qBAAAA,GAAAA,CAAAA,mBAAAA,GAAAA,SAAS,CAACC,QAAb,KAAG,IAAA,GAAA,KAAA,CAAA,GAAA,mBAAA,CAAoBnB,WAApB,EAAA,CAAkCoB,QAAlC,CAA2C,KAA3C,CAAH,KAAwD,IAAA,GAAA,qBAAA,GAAA,KAAA;GAEzD,GAAA,IAAA;AAAA,MADDC,KACC,GAAA,wBAAA,CAAA,IAAA,EAAA,SAAA,CAAA,CAAA;;EACJ,MAAMC,SAAS,GAAG,OAAON,QAAP,KAAoB,QAApB,GAA+B,CAACA,QAAD,CAA/B,GAA4CA,QAA9D,CAAA;AACA,EAAA,oBACIO,mBAAA,OAAA,EAAAC,cAAA,CAAA;AACIP,IAAAA,SAAS,EAAEQ,UAAU,CAAC,4BAAD,EAA+BR,SAA/B,EAA0C;MAC3D,mCAAqCV,EAAAA,KAAAA;KADpB,CAAA;AADzB,GAAA,EAIQc,KAJR,CAMKC,EAAAA,SAAS,CAACP,GAAV,CAAc,CAACT,QAAD,EAAWoB,CAAX,kBACXH,KAAC,CAAAI,aAAD,CAACJ,KAAK,CAACK,QAAP,EAAe;AAACrC,IAAAA,GAAG,EAAEmC,CAAAA;AAAN,GAAf,EACKA,CAAC,KAAK,CAAN,GAAU,IAAV,GAAiB,IADtB,eAEIH,KAAA,CAAAI,aAAA,CAAA,KAAA,EAAA,IAAA,EACKtB,SAAS,CAACC,QAAD,EAAWC,KAAX,EAAkBC,YAAlB,CAAT,CAAyCO,GAAzC,CAA6C,CAACxB,GAAD,EAAMsC,CAAN,kBAC1CN,KAAK,CAAAI,aAAL,CAAK,KAAL,EAAK;AAAApC,IAAAA,GAAG,EAAEsC,CAAAA;AAAL,GAAL,EAActC,GAAd,CADH,CADL,CAFJ,CADH,CANL,CADJ,CAAA;AAmBH;;;;"}