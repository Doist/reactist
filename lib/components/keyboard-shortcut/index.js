"use strict";Object.defineProperty(exports,"__esModule",{value:!0});var e=require("./keyboard-shortcut.js");exports.default=e.KeyboardShortcut;
//# sourceMappingURL=index.js.map
