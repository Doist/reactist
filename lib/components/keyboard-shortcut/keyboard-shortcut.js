"use strict";Object.defineProperty(exports,"__esModule",{value:!0});var e=require("../../_virtual/_rollupPluginBabelHelpers.js"),t=require("react"),r=require("classnames");function a(e){return e&&"object"==typeof e&&"default"in e?e:{default:e}}function c(e){if(e&&e.__esModule)return e;var t=Object.create(null);return e&&Object.keys(e).forEach((function(r){if("default"!==r){var a=Object.getOwnPropertyDescriptor(e,r);Object.defineProperty(t,r,a.get?a:{enumerable:!0,get:function(){return e[r]}})}})),t.default=e,t}var n=c(t),s=a(r);const l=["children","className","translateKey","isMac"];let o=e=>e;function u(e){switch(e.toLowerCase()){case"cmd":case"mod":return"⌘";case"control":case"ctrl":return"⌃";case"alt":return"⌥";case"shift":return"⇧";case"space":return"␣";default:return e}}function i(t){var r,a;let{children:c,className:i,translateKey:d=o,isMac:p=null!=(r=null==(a=navigator.platform)?void 0:a.toUpperCase().includes("MAC"))&&r}=t,f=e.objectWithoutProperties(t,l);const m="string"==typeof c?[c]:c;return n.createElement("span",e.objectSpread2({className:s.default("reactist_keyboard_shortcut",i,{"reactist_keyboard_shortcut--macos":p})},f),m.map((e,t)=>n.createElement(n.Fragment,{key:t},0===t?null:", ",n.createElement("kbd",null,function(e,t,r){const a=t?u:r,c=/\b(mod|cmd|ctrl|control|alt|shift)\b/i.test(e);return t||(e=e.replace(/\b(mod|cmd)\b/i,"ctrl")),e.split(/\s*\+\s*/).map((function(e){return function(e){return/^(mod|cmd|ctrl|control|alt|shift|space|super)$/i.test(e)}(e)?function(e){return e.charAt(0).toUpperCase()+e.slice(1).toLowerCase()}(a(e)):c&&1===e.length?e.toUpperCase():e}))}(e,p,d).map((e,t)=>n.createElement("kbd",{key:t},e))))))}i.setTranslateKey=e=>{o=e},exports.KeyboardShortcut=i;
//# sourceMappingURL=keyboard-shortcut.js.map
