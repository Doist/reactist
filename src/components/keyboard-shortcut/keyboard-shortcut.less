@import '../styles/constants.less';

.reactist_keyboard_shortcut {
    font-family: var(--reactist-font-family-monospace);
    font-size: 11px;

    > kbd > kbd {
        display: inline-block;
        padding: 3px 5px;
        margin: 0 2px;
        line-height: normal;
        line-height: 10px;
        color: @primary_font_color;
        vertical-align: middle;
        background-color: @color_grey_11;
        border: 1px solid @color_border_grey;
        border-radius: 6px;
        box-shadow: inset 0 -1px 0 @color_border_grey;
    }
}
