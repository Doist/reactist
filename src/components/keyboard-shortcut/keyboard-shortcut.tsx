import * as React from 'react'
import classNames from 'classnames'

//
// Support for setting up how to translate modifiers globally.
//

let globalTranslateKey = (key: string) => key

type TranslateKey = typeof globalTranslateKey

KeyboardShortcut.setTranslateKey = (tr: TranslateKey) => {
    globalTranslateKey = tr
}

function translateKeyMac(key: string) {
    switch (key.toLowerCase()) {
        case 'cmd':
        case 'mod':
            return '⌘'
        case 'control':
        case 'ctrl':
            return '⌃'
        case 'alt':
            return '⌥'
        case 'shift':
            return '⇧'
        case 'space':
            return '␣'
        default:
            return key
    }
}

//
// Some helpers
//

function capitalize(str: string) {
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()
}

function hasModifiers(str: string) {
    return /\b(mod|cmd|ctrl|control|alt|shift)\b/i.test(str)
}

function isSpecialKey(str: string) {
    return /^(mod|cmd|ctrl|control|alt|shift|space|super)$/i.test(str)
}

function parseKeys(shortcut: string, isMac: boolean, translateKey: TranslateKey) {
    const t = isMac ? translateKeyMac : translateKey
    const _hasModifiers = hasModifiers(shortcut)

    function mapIndividualKey(str: string) {
        if (isSpecialKey(str)) {
            return capitalize(t(str))
        }
        if (_hasModifiers && str.length === 1) {
            return str.toUpperCase()
        }
        return str
    }

    if (!isMac) {
        shortcut = shortcut.replace(/\b(mod|cmd)\b/i, 'ctrl')
    }

    return shortcut.split(/\s*\+\s*/).map(mapIndividualKey)
}

//
// The KeyboardShortcut component
//

type NativeSpanProps = React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLSpanElement>,
    HTMLSpanElement
>

type Props = Omit<NativeSpanProps, 'children'> & {
    /**
     * The shortcut to be represented as markup. It supports an intuitive syntax where you can
     * combine modifiers (cmd, ctrl, shift, alt) with single keys all concatenated with plus signs.
     * You can also pass various shortcuts as an array, which will be depicted separated by commas.
     */
    children: string | string[]
    /**
     * A function that allows you to change how some key names are represented. This may be useful,
     * for instance, to translate modifier names that are expressed differently in other languages
     * (e.g. `Ctrl` is named `Strg` in German).
     *
     * It defaults to a global version that leaves the key as is. You can pass your version as a
     * prop, or you can also set your own version of this global default one, so you don't need to
     * pass your own on each invocation of this component.
     *
     * ```js
     * import { KeyboardShortcut } from '@doist/reactist'
     * KeyboardShortcut.setTranslateKey = key => { ... }
     * ```
     *
     * Note: When the component detects the macOS operating system it bypasses key translation for
     * most modifiers and uses macOS-specific symbols. See the `isMac` prop for details.
     */
    translateKey?: TranslateKey
    /**
     * This prop is not meant to be passed. The component will automatically initialize it to `true`
     * if it detects that the current browser / operating system is on macOS, in which case modifier
     * keys are represented using macOS' notation (e.g. ⌘ ⌃ ⌥ ⇧).
     *
     * Though it is discouraged, if you don't want this special treatment in macOS, you can pass
     * `isMac={false}` in all invocations of this component.
     */
    isMac?: boolean
}

function KeyboardShortcut({
    children,
    className,
    translateKey = globalTranslateKey,
    isMac = navigator.platform?.toUpperCase().includes('MAC') ?? false,
    ...props
}: Props) {
    const shortcuts = typeof children === 'string' ? [children] : children
    return (
        <span
            className={classNames('reactist_keyboard_shortcut', className, {
                'reactist_keyboard_shortcut--macos': isMac,
            })}
            {...props}
        >
            {shortcuts.map((shortcut, i) => (
                <React.Fragment key={i}>
                    {i === 0 ? null : ', '}
                    <kbd>
                        {parseKeys(shortcut, isMac, translateKey).map((key, j) => (
                            <kbd key={j}>{key}</kbd>
                        ))}
                    </kbd>
                </React.Fragment>
            ))}
        </span>
    )
}

export { KeyboardShortcut }
