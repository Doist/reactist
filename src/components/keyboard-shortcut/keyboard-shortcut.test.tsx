import * as React from 'react'
import { render } from '@testing-library/react'
import { KeyboardShortcut } from './keyboard-shortcut'

/**
 * Renders the KeyboardShortcut component twice, once for macOS and non-macOS setting.
 * @param shortcut the keyboard shortcut string
 * @param props extra props for the KeyboardShortcut component
 */
function renderKeyboardShortcut(
    shortcut: string | string[],
    props: Omit<React.ComponentProps<typeof KeyboardShortcut>, 'children' | 'isMac'> = {},
) {
    const { container } = render(
        <>
            <KeyboardShortcut {...props} isMac={false}>
                {shortcut}
            </KeyboardShortcut>
            <KeyboardShortcut {...props} isMac={true}>
                {shortcut}
            </KeyboardShortcut>
        </>,
    )
    return container
}

describe('KeyboardShortcut', () => {
    it('works as expected', () => {
        const element = renderKeyboardShortcut('ctrl + alt + shift + e')
        expect(element).toMatchSnapshot()
    })

    it('recognizes mod/cmd as a modifier that behaves differently in macOS and outside macOS', () => {
        const element = renderKeyboardShortcut(['mod + k', 'cmd+j'])
        expect(element).toMatchSnapshot()
    })

    it('supports passing various shortcuts, which it will separate with commas', () => {
        const element = renderKeyboardShortcut(['p1', 'p2'])
        expect(element).toMatchSnapshot()
    })

    it('only capitalizes a non-modifier part if the overall shortcut has modifiers', () => {
        const element = renderKeyboardShortcut(['e', 'alt+e'])
        expect(element).toMatchSnapshot()
    })

    it('allows to customize how modifiers are translated', () => {
        const element = renderKeyboardShortcut('ctrl + alt + shift + e', {
            translateKey: (modifier) => (modifier.length === 1 ? modifier : `[${modifier}]`),
        })
        expect(element).toMatchSnapshot()
    })

    it('recognizes "Super" as a special key to be translated', () => {
        const element = renderKeyboardShortcut('super + q', {
            translateKey: (modifier) => `[${modifier}]`,
        })
        expect(element).toMatchSnapshot()
    })
})
