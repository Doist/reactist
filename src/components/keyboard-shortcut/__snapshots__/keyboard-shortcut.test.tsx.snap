// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardShortcut allows to customize how modifiers are translated 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        [ctrl]
      </kbd>
      <kbd>
        [alt]
      </kbd>
      <kbd>
        [shift]
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        ⌃
      </kbd>
      <kbd>
        ⌥
      </kbd>
      <kbd>
        ⇧
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
</div>
`;

exports[`KeyboardShortcut only capitalizes a non-modifier part if the overall shortcut has modifiers 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        e
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        Alt
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        e
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        ⌥
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
</div>
`;

exports[`KeyboardShortcut recognizes "Super" as a special key to be translated 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        [super]
      </kbd>
      <kbd>
        q
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        Super
      </kbd>
      <kbd>
        q
      </kbd>
    </kbd>
  </span>
</div>
`;

exports[`KeyboardShortcut recognizes mod/cmd as a modifier that behaves differently in macOS and outside macOS 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        Ctrl
      </kbd>
      <kbd>
        K
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        Ctrl
      </kbd>
      <kbd>
        J
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        ⌘
      </kbd>
      <kbd>
        K
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        ⌘
      </kbd>
      <kbd>
        J
      </kbd>
    </kbd>
  </span>
</div>
`;

exports[`KeyboardShortcut supports passing various shortcuts, which it will separate with commas 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        p1
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        p2
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        p1
      </kbd>
    </kbd>
    , 
    <kbd>
      <kbd>
        p2
      </kbd>
    </kbd>
  </span>
</div>
`;

exports[`KeyboardShortcut works as expected 1`] = `
<div>
  <span
    class="reactist_keyboard_shortcut"
  >
    <kbd>
      <kbd>
        Ctrl
      </kbd>
      <kbd>
        Alt
      </kbd>
      <kbd>
        Shift
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
  <span
    class="reactist_keyboard_shortcut reactist_keyboard_shortcut--macos"
  >
    <kbd>
      <kbd>
        ⌃
      </kbd>
      <kbd>
        ⌥
      </kbd>
      <kbd>
        ⇧
      </kbd>
      <kbd>
        E
      </kbd>
    </kbd>
  </span>
</div>
`;
